waiter/integration/waiter/token_watch_integration_test.clj

```diff
@@ -0,0 +1,362 @@
+
+      (testing "hard deleting a token reflects change in token-watch-state"
+        (let [_ (delete-token-and-assert waiter-url token)
+              goal-fn (fn [{:keys [body] :as response}]
+                        (let [{:strs [state]} (try-parse-json body)]
+                          (assert-response-status response 200)
+                          (nil? (get-in state ["token->index" token]))))]
+          (is (await-goal-response-for-all-routers goal-fn watch-state-request-fn router_urls)))))))
+
+(defn- start-watch
+  [router-url cookies & {:keys [query-params] :or {query-params {"include" ["metadata", "deleted"]
+                                                                 "watch" "true"}}}]
+  (let [{:keys [body error]} (make-request router-url "/tokens"
+                                           :cookies cookies
+                                           :fold-chunked-response? false
+                                           :query-params query-params)
+        token->index-atom (atom {})
+        query-state-fn (fn [] @token->index-atom)
+        exit-chan (async/promise-chan)
+        go-chan
+        (async/go-loop [token->index @token->index-atom]
+          (reset! token->index-atom token->index)
+          (let [[msg chan] (async/alts! [body exit-chan] :priority true)
+                next-token->index
+                (condp = chan
+                  body
+                  (when (some? msg)
+                    (let [{:strs [object type]} (try-parse-json msg)]
+                      (case type
+                        "INITIAL"
+                        (reduce
+                          (fn [token->index index]
+                            (assoc token->index (get index "token") index))
+                          {}
+                          object)
+
+                        "EVENTS"
+                        (reduce
+                          (fn [token->index {:strs [object type]}]
+                            (case type
+                              "UPDATE"
+                              (assoc token->index (get object "token") object)
+                              "DELETE"
+                              (dissoc token->index (get object "token") object)
+                              (throw (ex-info "Unknown event type received in EVENTS object" {:event object}))))
+                          token->index
+                          object)
+                        (throw (ex-info "Unknown event type received from watch" {:event msg})))))
+
+                  exit-chan
+                  (do
+                    (async/close! body)
+                    false))]
+            (when next-token->index
+              (recur next-token->index))))]
+    {:exit-chan exit-chan
+     :error-chan error
+     :go-chan go-chan
+     :query-state-fn query-state-fn}))
+
+(defn- start-watches
+  [router-urls cookies]
+  (mapv
+    #(start-watch % cookies)
+    router-urls))
```

Comment:
If we maintain a router-id/url->watch map, we can provide more explanations in the failed assertions about what failed and on which router.